### PR TITLE
Hook up Registration v2

### PR DESCRIPTION
**Pull request by hanyulo**
## Features
1. Registration System 
2. Bookmark 

## Targeted PR Branch
twreporter: registration (from twreporter: master)

## Related Repos
• [twreporter-react-components](https://github.com/twreporter/twreporter-react-components)
• [twreporter-redux](https://github.com/twreporter/twreporter-redux)
• [twreporter-registration](https://github.com/twreporter/react-redux-registration)
• [twreporter-react](https://github.com/twreporter/twreporter-react)

## Reviewer 
@nickhsine

### CODE REVIEW

File: src/containers/BookmarkList.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+  _getNextPage() {
+    const offset = this.bookmarkData.length
+    this.props.getBookmarks(offset, DEFAULT_LIMIT, SORT)
+      .then(({ data, total }) => {
+        this.total = total
+        if (this.bookmarkData.length <= total) {
+          data.forEach((bookmark) => {
+            this.bookmarkData.push(bookmark)
+          })
```
Comment (nickhsine) [inline src/containers/BookmarkList.js]: `this.bookmarkData = this.bookmarkData.concact(data)`

File: src/containers/BookmarkList.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          this.forceUpdate()
```
Comment (nickhsine) [inline src/containers/BookmarkList.js]: move `this.bookmarkData` to `this.state.bookmarkData`?

In this component, changes of bookmarkData(add/delete) will affect how this component renders. Hence, bookmarkData should be the state of this component.

BTW, you could move total into bookmarkData state.

File: src/containers/SignIn.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          <FacebookButton />
+          <GoogleButton />
+        </SignInForm>
+      </Container>
+    )
+  }
+}
+
+
+export default connect(null, { setHeaderInfo })(SignIn)
```
Comment (nickhsine) [inline src/containers/SignIn.js]: Isn't [SignIn/SignUp](https://github.com/twreporter/react-redux-registration/blob/b3d885e7a545700240509e3ff3b13aa9bbf78190/src/containers/sign-in-prototype.js#L230) already hooked with redux.connect ?

File: src/index.js [hunk 4]
```diff
@@ -31,6 +32,25 @@ const store = configureStore(reduxState)
...
+  store.dispatch(deletAuthInfoAction())
+}
+
+// Check if token existed in localStorage and expired
+// following preocedure is for both accoutn and oAuth SignIn
+// 7 = 7 days
+store.dispatch(authUserByTokenAction(7, auth.authType))
+  .then(() => {})
+  .catch(() => {})
+
```
Comment (nickhsine) [inline src/index.js]: Do `authenticateUserByToken || authUserByTokenAction` need to be a `Promise`? 
Are [those actions](https://github.com/twreporter/react-redux-registration/blob/b3d885e7a545700240509e3ff3b13aa9bbf78190/src/actions/authentication.js#L137-L138) you send async functions?
If not async functions, you don't need to make it a Promise.

File: src/reducers/index.js [hunk 5]
```diff
@@ -10,6 +11,21 @@ import twreporterRedux from 'twreporter-redux'
...
+  apiUrl: '',
+  signUp: '',
+  signIn: '',
+  activate: '',
+  oAuthProviders: {
+    google: '',
+    facebook: ''
+  },
+  location: 'http://testtest.twreporter.org:3000',
+  domain: 'twreporter.org'
```
Comment (nickhsine) [inline src/reducers/index.js]: How to configure it ?
If you want different values in different environments.

File: src/routes/index.js [hunk 6]
```diff
@@ -138,6 +138,44 @@ export default function (history = browserHistory) {
...
+          path="activate"
+          getComponent={(nextStateWithLocation, cb) => {
+            require.ensure([], (require) => {
+              const module = require('../containers/Activation')
+              cb(null, module.default || module)
+            })
+          }}
+        />
+        <Route
+          path="signin(/:type/:slug)"
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/routes/index.js]: I set type and slug to be optional. User can just surf to sign in page or be redirected by the app, which will carry the type and slug info.

File: src/utils/style-utils.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    @media (max-width: ${breakPoints.desktopMaxWidth}) {
+      ${css(...args)}
+    }
+  `,
+  overDesktop: (...args) => css`
+    @media (min-width: ${breakPoints.overDesktopMinWidth}) {
+      ${css(...args)}
+    }
+  `
+}
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/utils/style-utils.js]: We may need it in future to be consistent with index page outline ?

File: src/reducers/index.js (hunk 5 above)
Comment (hanyulo) [reply src/reducers/index.js]: What do mean by the word environment ? You mean without redux ? 

In the case, I configure it in server.js file

File: server/config.js [hunk 8]
```diff
@@ -17,5 +17,17 @@ module.exports = Object.assign({
...
+    apiUrl: 'http://localhost:8080',
+    signUp: '/v1/signup',
+    signIn: '/v1/login',
+    activate: '/v1/activate',
+    oAuthProviders: {
+      google: '/v1/auth/google',
+      facebook: '/v1/auth/facebook'
+    },
+    location: 'http://testtest.twreporter.org:3000',
+    domain: 'twreporter.org'
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply server/config.js]: 👍

File: src/containers/Activation.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    <ActivePage
+      activateRedirectPath={'/'}
+      browserHistory={browserHistory}
+      {...props}
+    />
+  </div>
+)
+
+
+export default connect()(Activation)
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/containers/Activation.js]: 👍

File: src/containers/Article.js [hunk 10]
```diff
@@ -127,10 +130,13 @@ class Article extends PureComponent {
...
     this._onResize =  _.throttle(this._onResize, 500).bind(this)
     this.toggleTools = this._toggleTools.bind(this)
+    this.handleOnClickBookmark = this._handleOnClickBookmark.bind(this)
 
     this.state = {
       fontSize:'medium',
-      isFontSizeSet:false
+      isFontSizeSet:false,
+      bookmarkListed: false,
+      bookmarkId: DEFAULT_INVALID_BOOKMARK_ID
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/containers/Article.js]: 👍   change to isBookmarkListed

File: src/containers/BookmarkList.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    this.total = 0
+    this.targetSlug = ''
+  }
+
+  componentWillMount() {
+    this.getNextPage()
+    const { setHeaderInfo } = this.props
+    setHeaderInfo({
+      pageTheme: LIGHT,
+      pageType: AUTHOR_PAGE
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/containers/BookmarkList.js]: change to BOOKMARK_LIST

File: src/containers/SignIn.js (hunk 3 above)
Comment (hanyulo) [reply src/containers/SignIn.js]: SignUp to a simple component. The SignIn fille need to dispatch setHeaderInfo, so I keep it as a container.

File: src/index.js (hunk 4 above)
Comment (hanyulo) [reply src/index.js]: 👍

File: src/reducers/index.js (hunk 5 above)
Comment (hanyulo) [reply src/reducers/index.js]: So I update the config file. The config file contains the registration default value and the value set are created based on different env (dev/production).

File: src/components/article/tools/DesktopArticleTools.js [hunk 12]
```diff
@@ -1,9 +1,30 @@
...
-import { Link } from 'react-router'
-import styles from './DesktopArticleTools.scss'
 import BackToTopicIcon from '../../../../static/asset/article-back-to-topic.svg'
+import styles from './DesktopArticleTools.scss'
+import styled from 'styled-components'
+import BookmarkAddedIcon from '../../../../static/asset/added-bookmark-desktop.svg'
+import BookmarkUnaddedIcon from '../../../../static/asset/add-bookmark-desktop.svg'
 import { CSSTransitionGroup } from 'react-transition-group'
 import { LINK_PREFIX } from '../../../constants/link-prefix'
+import { Link } from 'react-router'
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/components/article/tools/DesktopArticleTools.js]: 👍

**Comment by nickhsine** [comment]
Remember to rebase your commits.

**Pull request merged**